src/Payum/Action/NotifyAction.php

```diff
@@ -19,78 +20,120 @@
-        $payment->setDetails($details->getArrayCopy());
-
-        // If payment hasn't been validated yet, validate its status against Alma's payment state
-        if (in_array($payment->getState(), [PaymentInterface::STATE_NEW, PaymentInterface::STATE_PROCESSING], true)) {
-            try {
-                $this->gateway->execute(new ValidatePayment($payment));
-            } catch (\Exception $e) {
-                $error = [
-                    "error" => true,
-                    "message" => $e->getMessage()
-                ];
-
-                throw new HttpResponse(
-                    json_encode($error),
-                    Response::HTTP_INTERNAL_SERVER_ERROR,
-                    ["content-type" => "application/json"]
-                );
-            }
-        }
-
-        // $details is the request's model here, but we used a copy above passed down the ValidatePaymentAction through
-        // the $payment object, which is the model of our own request.
-        // Since Payum will use whatever is in the original details model to overwrite the payment's details data, we
-        // need to make sure we copy everything that was set on the payment itself back to the NotifyRequest model.
-        $details->replace($payment->getDetails());
-
-        // Down here means the callback has been correctly handled, regardless of the final payment state
-        throw new HttpResponse(
-            json_encode(["success" => true, "state" => $payment->getDetails()[AlmaBridgeInterface::DETAILS_KEY_IS_VALID]]),
-            Response::HTTP_OK,
-            ["content-type" => "application/json"]
-        );
-    }
-
-    public function supports($request): bool
-    {
-        return $request instanceof Notify
-            && $request->getModel() instanceof ArrayAccess
-            && $request->getFirstModel() instanceof PaymentInterface;
-    }
+	use GatewayAwareTrait;
+	use ApiAwareTrait;
+
+	/**
+	 * @var AlmaBridgeInterface
+	 */
+	protected $api;
+
+	/**
+	 * @var RequestStack
+	 */
+	protected $requestStack;
+
+	/**
+	 * @var SecurityHelper
+	 */
+	protected $securityHelper;
+
+	public function __construct(RequestStack $requestStack)
+	{
+		$this->apiClass = AlmaBridge::class;
+		$this->requestStack = $requestStack;
+		$this->securityHelper = new SecurityHelper();
+	}
+
+	/**
+	 * @param Notify $request
+	 */
+	public function execute($request): void
+	{
+		RequestNotSupportedException::assertSupports($this, $request);
+		$httpRequest = new GetHttpRequest();
+		$this->gateway->execute($httpRequest);
+
+		/** @var string $signature */
+		$signature = ArrayObject::ensureArrayObject($httpRequest->headers)->get('x-alma-signature')[0];
+		$query = ArrayObject::ensureArrayObject($httpRequest->query);
+
+		/** @var string $payment_id */
+		$payment_id = $query->get(AlmaBridgeInterface::QUERY_PARAM_PID);
+
+		/** @var PaymentInterface $payment */
+		$payment = $request->getFirstModel();
+
+		try {
+			$this->securityHelper->isHmacValidated(
```

Comment:
The HMAC verification didn't work, I think we need to explore how does work the NotifyAction and tokenFactory

Reply:
<img width="854" alt="Capture d’écran 2024-10-03 à 16 34 44" src="https://github.com/user-attachments/assets/bfc88111-6457-4ecd-9a08-9f60d964eb1a">